subsys/net/lib/coap/coap.c

```diff
@@ -0,0 +1,1449 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_COAP)
+#define SYS_LOG_DOMAIN "coap"
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <stdlib.h>
+#include <stddef.h>
+#include <zephyr/types.h>
+#include <string.h>
+#include <stdbool.h>
+#include <errno.h>
+
+#include <misc/byteorder.h>
+#include <net/buf.h>
+#include <net/net_pkt.h>
+#include <net/net_ip.h>
+
+#include <net/coap.h>
+
+struct option_context {
+	u16_t delta;
+	u16_t offset;
+	struct net_buf *frag;
+};
+
+#define COAP_VERSION 1
+
+#define COAP_MARKER 0xFF
+
+#define BASIC_HEADER_SIZE 4
+
+#define PKT_WAIT_TIME K_SECONDS(1)
+
+static u8_t option_header_get_delta(u8_t opt)
+{
+	return (opt & 0xF0) >> 4;
+}
+
+static u8_t option_header_get_len(u8_t opt)
+{
+	return opt & 0x0F;
+}
+
+static void option_header_set_delta(u8_t *opt, u8_t delta)
+{
+	*opt = (delta & 0xF) << 4;
+}
+
+static void option_header_set_len(u8_t *opt, u8_t len)
+{
+	*opt |= (len & 0xF);
+}
+
+static int get_coap_packet_len(struct net_pkt *pkt)
+{
+	struct net_buf *frag;
+	u16_t offset;
+	u16_t len;
+
+	frag = net_frag_read_be16(pkt->frags,
+				  net_pkt_ip_hdr_len(pkt) +
+				  net_pkt_ipv6_ext_len(pkt) +
+				  4, &offset, &len);
+	if (!frag && offset == 0xffff) {
+		return -1;
+	}
+
+	return len - NET_UDPH_LEN;
+}
+
+static int decode_delta(u16_t num, struct option_context *context,
+			u16_t *decoded)
+{
+	int hdrlen = 0;
+
+	if (num == 13) {
+		u8_t val;
+
+		context->frag = net_frag_read_u8(context->frag, context->offset,
+						 &context->offset, &val);
+		if (!context->frag && context->offset == 0xffff) {
+			return -EINVAL;
+		}
+
+		num = val + 13;
+		hdrlen += 1;
+	} else if (num == 14) {
+		u16_t val;
+
+		context->frag = net_frag_read_be16(context->frag,
+						   context->offset,
+						   &context->offset, &val);
+		if (!context->frag && context->offset == 0xffff) {
+			return -EINVAL;
+		}
+
+		num = val + 269;
+		hdrlen += 2;
+	} else if (num == 15) {
+		return -EINVAL;
+	}
+
+	*decoded = num;
+
+	return hdrlen;
+}
+
+static int parse_option(const struct coap_packet *cpkt,
+			struct option_context *context,
+			struct coap_option *option)
+{
+	u16_t opt_len;
+	u16_t delta;
+	u16_t len;
+	u8_t opt;
+	int r;
+
+	context->frag = net_frag_read_u8(context->frag, context->offset,
+					 &context->offset, &opt);
+	if (!context->frag && context->offset == 0xffff) {
+		return -EINVAL;
+	}
+
+	opt_len = 1;
+
+	/* This indicates that options have ended */
+	if (opt == COAP_MARKER) {
```

Comment:
I'm hitting a case where context->offset == 0 here.  When this happens the incoming packet has no payload and no options (it's a response packet).  What happens is there's no COAP_MARKER and it proceeds to generate EINVAL below.

IMHO the logic should be:
/* This indicates that we hit the end of the packet data or options have ended */
if (context->offset == 0 || opt == COAP_MARKER) {

Reply:
Actually it should it here in parse_options() while {}
                if (opt_len >= coap_len) {
                        break;
                }
I hope I have handled if the packet doesn't have COAP_MARKER or even zero Options. In your case coap_len (UDP payload[coap packet len] - coap headers len] should be 0, if not something wrong.